### ISSUE TITLE
V2 Subscriptions

### ISSUE THREAD

**Issue by jorgebucaran** [jorgebucaran/hyperapp#752]
## Summary

Working with traditional event emitters requires a lot of complicated resource management like adding and removing event listeners, closing connections, clearing intervals—not to mention testing asynchronous code is hard. What happens when the source you are subscribed to shuts down? How do you cancel or restart a subscription?

Subscriptions are a declarative abstraction layer for [managing global events](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers) like window events and custom event streams such as clock ticks, [geolocation](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation) changes, handling [push notifications](https://developers.google.com/web/fundamentals/push-notifications/handling-messages), and [WebSockets](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API) in the browser.

Think of subscriptions as "virtual DOM meets event streams". Events are represented using plain objects called _subscriptions_. When the program state changes we compare the new subscriptions against the old, compute their differences and rewire the underlying connections to match the desired state.

In this issue, I'll introduce everything you need to know about the upcoming Subscriptions API.

## Background

Let's start with the good news. You already know how subscriptions work! Here is one that listens to button clicks. As long as the button remains in the view, Hyperapp will dispatch an Increment action every time the button is clicked.

```jsx
const Increment = state => state + 1

app({
  init: 0,
  view: state => <button onClick={Increment}>Clicked {state} time(s)</button>,
  container: document.body
})
```

In the next example, a Reset button is shown only after the Increment button is clicked at least one time. Click it to set the count back to 0 and make the button disappear. The declarative view model allows us to add and remove elements dynamically, which in turn cancels any event listeners it was subscribed to.

```jsx
const Reset = () => 0
const Increment = state => state + 1

app({
  init: Reset,
  view: state => (
    <main>
      <h1>{state}</h1>
      <button onClick={Increment}>Clicked {state} time(s)</button>
      {state > 0 && <button onClick={Reset}>Reset</button>}
    </main>
  ),
  container: document.body
})
```

The take-away here is that event listeners and subscriptions are one and the same.

But what about _global_ event listeners like window, mouse and keyboard events? Traditionally, we would use the imperative DOM addEventListener and removeEventListener API to subscribe to an event stream. The next example tracks the mouse position as you move it on the screen. Click the toggle button to un/subscribe from the event.

```jsx
const toggleEventListener = (event, eventListener, shouldSubscribe) => {
  if (shouldSubscribe) {
    addEventListener(event, eventListener)
  } else {
    removeEventListener(event, eventListener)
  }
}

const MouseMoved = ({ x, y }) => {
  text.textContent = `${x},${y}`
}

toggleEventListener("mousemove", MouseMoved, true)

const text = document.body.appendChild(document.createElement("h1"))
const button = document.body.appendChild(document.createElement("button"))

let isTracking = true
button.textContent = "Toggle Subscription"
button.onclick = () => {
  toggleEventListener("mousemove", MouseMoved, (isTracking = !isTracking))
}
```

The Subscriptions API allow us to do the same work in a declarative fashion.

```jsx
import { h, app } from "hyperapp"
import * as Mouse from "@hyperapp/mouse"

const Toggle = state => ({ ...state, isTracking: !state.isTracking })
const MouseMoved = (state, { x, y }) => ({ ...state, position: `${x},${y}` })

app({
  init: {
    position: "",
    isTracking: true
  },
  view: state => (
    <main>
      <h1>{state.position}</h1>
      <button onClick={Toggle}>Toggle Subscription</button>
    </main>
  ),
  subscriptions: state =>
    state.isTracking && Mouse.moves({ action: MouseMoved }),
  container: document.body
})
```

## Declarative subscriptions

We want to be notified when something happens in the outside world: mouse movements, keyboard presses, clock ticks, browser repaints, and so on. Subscriptions allow us to listen for such things.

A subscription is a plain object that identifies the type of the event you want to listen to, the action to dispatch with the event data and other data it needs. Not every subscription produces an [Event](https://developer.mozilla.org/en-US/docs/Web/API/Event). They can be used to model other event streams such as [geolocation](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation/watchPosition) changes, [time intervals](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setInterval) and [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent).

To declare subscriptions we'll use the new `subscriptions` function. It receives the current state and returns one or more subscriptions. Like with effects (#750) you'll likely be using an existing subscription library to create subscriptions: **@hyperapp/time**, **@hyperapp/mouse**, **@hyperapp/keyboard**, etc.

```jsx
import { app } from "hyperapp"
import * as Time from "@hyperapp/time"
import { Tick } from "./actions"

app({
  subscriptions: state => Time.tick({ interval: 1000, action: Tick })
})
```

Hyperapp will call this function to refresh your subscriptions every time the state changes. There you can conditionally add, update and remove subscriptions much the same way you do with elements in the view function.

###### Vanilla

```jsx
export const subscriptions = state => [
  state.isIntro || state.isLost || state.isWon
    ? MouseClicks(startGame)
    : state.isInPlay && [
        AnimationFrame(tick),
        KeyDowns(keyPressed),
        KeyUps(move)
      ]
]
```

###### JSX

```jsx
export const subscriptions = state => [
  state.isIntro || state.isLost || state.isWon ? (
    <Mouse.clicks action={StartGame} />
  ) : (
    state.isInPlay && [
      <Animation.frame action={Tick} />,
      <Key.downs action={keyPressed} />,
      <Key.ups action={move} />
    ]
  )
]
```

If a new subscription appears in the array, we'll _start_ the subscription. When a subscription leaves the array, we'll _cancel_ it. If any of its properties change, we'll _restart_ it. Otherwise, we'll skip to the next subscription until we've seen them all.

## Examples

<a name="svg-clock"></a>

### SVG Clock

```jsx
const angle = t => (2 * Math.PI * t) / 60000
const Tick = (state, data) => data.time

app({
  init: Date.now(),
  view: state => (
    <svg viewBox="0 0 100 100" width="300px">
      <circle cx="50" cy="50" r="45" fill="#00caff" />
      <line
        x1="50"
        y1="50"
        x2={50 + 40 * Math.cos(angle(state))}
        y2={50 + 40 * Math.sin(angle(state))}
        stroke="#111"
      />
    </svg>
  ),
  subscriptions: state => <Time.tick interval={1000} action={Tick} />,
  container: document.body
})
```

<a name="countdown-clock"></a>

### Countdown Clock

```jsx
import { h, app } from "hyperapp"
import * as Time from "@hyperapp/time"

const pad = n => (n < 10 ? "0" + n : n)

const humanizeTime = t => {
  const hours = (t / 3600) >> 0
  const minutes = ((t - hours * 3600) / 60) >> 0
  const seconds = (t - hours * 3600 - minutes * 60) >> 0
  return `${pad(minutes)}:${pad(seconds)}`
}

const initialState = {
  count: 5, // Seconds
  paused: true
}

const Reset = state => initialState
const Toggle = state => ({ ...state, paused: !state.paused })
const Elapse = state =>
  state.count === 0
    ? initialState
    : {
        ...state,
        count: state.count - 1
      }

app({
  init: initialState,
  view: state => (
    <main>
      <h1>{humanizeTime(state.count)}</h1>
      <button onClick={Toggle}>{state.paused ? "Start" : "Paused"}</button>
      <button onClick={Reset}>Reset</button>
    </main>
  ),
  subscriptions: state => [
    !state.paused && <Time.tick interval={1000} action={Elapse} />
  ],
  container: document.body
})
```

## Implementing subscriptions

If you want to implement your own subscription then either one (or more) of these things are happening:

1.  Hyperapp's core subscription libraries lacks an important subscription. _Solution_: Request it.
2.  You've created a new service `PhaserLink` and there is no Hyperapp subscription for it. _Solution_: You'll need to write one.
3.  You want to write your own subscription. Period. _Solution_: Sure, why not?

Let's implement a few subscriptions to see how they work. There are two parts to implementing your own subscription:

- The _subscription constructor_ takes an object with the properties required by the subscription and returns an object that Hyperapp can make sense of.

- The _effect runner_ encapsulates the subscription setup: adding event listeners, installing a callback, etc. It receives what the subscription constructor returns and is responsible for dispatching actions through Hyperapp's dispatch function. It must return a function capable of unsubscribing from the event so that Hyperapp knows how to cancel the subscription if it needs to.

> See also [#750/implementing-effects](https://github.com/hyperapp/hyperapp/issues/750#implementing-custom-effects)

### Time ticks

```jsx
const effect = function(props, dispatch) {
  const id = setInterval(
    () => dispatch(props.action, { time: Date.now() }),
    props.interval
  )
  return () => clearInterval(id)
}

export const tick = props => ({
  effect: effect,
  action: props.action,
  interval: props.interval
})
```

### Mouse moves

```jsx
const effect = (props, dispatch) => {
  const eventListener = event => dispatch(props.action, event)
  addEventListener("mousemove", eventListener)
  return () => removeEventListener("mousemove", eventListener)
}

export const moves = props => ({
  effect: effect,
  action: props.action
})
```

## How does this all work?

Hyperapp calls the subscriptions function to create a new batch of _subscriptions_ every time the state changes. This is analogous to how the view function is called to create a new virtual node tree.

The next step is traversing the batch and compare each subscription with the ones in the old batch positionally. To determine if two subscriptions are equivalent, we diff the subscriptions object. If any property checks false for strict equality, we'll restart the subscription.

```jsx
for (const { subscription, oldSubscription } in batch)
  if (subscription)
    if (oldSubscription) {
      if (changed(subscription, oldSubscription)) restart(subscription)
    } else start(subscription)
  else if (oldSubscription) cancel(oldSubscription)
```

**Comment by selfup** [comment]
Excited to see this. I have this idea for using Effects/Subs for a game loops and theoretically it should work quite well with canvas 😂

**Comment by okwolf** [comment]
@selfup I have a subscription that implements an animation loop using `requestAnimationFrame`: https://github.com/hyperapp/fx/blob/HAv2/src/subs/Animation.js

**Comment by selfup** [comment]
Oh sweet! Glad I am not alone in those thoughts. I am tethering right now (travel) so I'll check it out tomorrow 🎉

**Comment by okwolf** [comment]
If `subscriptions` are still the [preferred option for interop](https://github.com/hyperapp/hyperapp/pull/726/files#r201544084), we need to document that.

**Comment by jorgebucaran** [comment]
@okwolf Good point. Subscriptions are the preferred option for interoperability now. In fact, I am not sure about returning the internal `dispatch` function either, so there's a vacancy open for returning something and I am taking ideas.

**Comment by selfup** [comment]
I would love for the internal dispatch to be exposed, but I can understand if you don't want to. Sometimes it's just easier to just call dispatch directly for that legacy code that just won't go away 😂 

Thanks for bringing it up!

@okwolf seems that subs look like effects after all? The constructors at least. Anyways, still stoked to see what we discover as we play around with effects/subs/action for v2 🎉

**Comment by jorgebucaran** [comment]
I think it's a wrap. https://github.com/hyperapp/hyperapp/issues/752

**Comment by SteveALee** [comment]
Just a thought but are subscription librarues and effects libraries separate or would you combine them into drivers as cycle does? 

That then makes me wonder about async if http request + response would be effect + event. Again cycle does it that way but you then need IDs to tie the 2 together.

**Comment by jorgebucaran** [comment]
@SteveALee I'd like to group libraries by a common domain, e.g., **@hyperapp/time** should export both a `delay` effect and a `tick` subscription (names are still provisional at this moment). 

**@hyperapp/http** should export a `fetch` effect that wraps the Fetch API (or xmlhttprequest). I don't know Cycle.js, but there will be no need for IDs. I have no idea what those would be used for.

Here is an example.

```jsx
const Increment = (state, value) => state + value
const DelayIncrement = (state, { value, duration }) => [
  state,
  delay({ action: [Increment, value], duration })
]

app({
  init: 0,
  view: state => (
    <main>
      <h1>{state}</h1>
      <button onClick={[Increment, 2]}>Add 2 Now</button>
      <button onClick={[DelayIncrement, { value: 100, duration: 5000 /*ms*/ }]}>
        Add 100; 5s Later
      </button>
    </main>
  ),
  subscriptions: state => tick({ interval: 1000, action: [Increment, 1] }),
  container: document.body
})
```

![2018-09-03 18 57 48](https://user-images.githubusercontent.com/56996/44980622-5f6c4680-afab-11e8-89ae-de798fea3092.gif)

---

One more thing, aborting a fetch request is not possible with an effect. Effects model side-effects. Subscriptions model event streams. So, I imagine there could be a fetch subscription (subject to a better name) that lets you create or tap into a transfer stream which can be canceled at any time.

**Comment by SteveALee** [comment]
@jorgebucaran 

> I'd like to group libraries by a common domain, e.g., @hyperapp/time should export both a delay effect and a tick subscription

Great - I think that's is best

The example is great.  I was wondering if I might group my actions into a object as V1 does. Purely for stylistic reasons. Was there a reason you do NOT do that? other and needing ref with `actions.Increment`

Sorry for confusion re fetch. As long a Fetch effect is async that's all good (actually I need to see examples of async effects, I assume you have them). I've worked with many async comms protocols in systems dev and was really complicating the issues

So an alternative style of handling Async remote request / responses rather than effect returning a promise that 'completes' on the expected response is to completely decouple them. 

1) Generate the effect which sends the request 
2) Receive a event some time later when the response arrives (or not) 

There is no wait of promise at all. The event just comes in. You can fire off multiple effects and then handle the events as the arrive (in any order). However you usually need to tie them together in the correct sequence (, like TCP/IP does it's packet matching), depending on the request / response protocol (perhaps  multiple responses are possible for one request). So either the protocol lets you supply an ID in the request and returns it in the matching response. Or you need to handle them locally for the effect and event.

Cycle takes this approach, leaving user-land to manage the IDs

Anyway with HTTP Req/Resp that is not necessary  ASFAIK :) 

So ignore me

**Comment by SteveALee** [comment]
> `onClick={[DelayIncrement, { value: 100, duration: 5000 /*ms*/ }]}`

yay! Tuples :)

**Comment by guido4000** [comment]
This seems to be pretty similar to Reason subscriptions, which are very helpful.

https://github.com/reasonml/reason-react/blob/master/docs/subscriptions-helper.md

**Comment by jorgebucaran** [comment]
Fascinating. I'm not familiar with Reason subscriptions, but you must be right @guido4000. I have mixed feelings about their API, and I am not sure where the similarities begin and end, but we're trying to accomplish the same thing and that's what matters.

V2 subscriptions are based on elm's subscriptions (and tasks). Like elm, each subscription can be added and removed conditionally, but  we can express them using an array right off the bat (where as in elm you need to use Sub.batch). Unlike elm, deeply nested batch of subscriptions can be expressed using nested arrays and any falsy elements in the array denote a subscription should be canceled. We're holding firm on the functional programming front, but also embracing JavaScript dynamic nature.

Users will likely be using existing subscriptions libraries, **@hyperapp/geolocation**, etc., instead of rolling their own, but I wanted  our subscriptions to be more accessible than elm's subscriptions, even to regular users, so I designed the API as a user-facing API and not a low-level API. When I was learning elm subscriptions I struggled, not to understand the concept (well I didn't for a while), but to implement my own custom subscriptions. In V2 this should be a piece of cake. 🍰

**Comment by okwolf** [comment]
So for the purposes of interoperability, do we have any best practices to recommend?

Would it make sense to have a subscription that adds a listener for [`CustomEvent`s](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent) and then dispatch those from outside of Hyperapp?

Or would it be better to use some sort of small event bus?

Perhaps some other approach I haven't considered yet?

**Comment by sergey-shpak** [comment]
@okwolf @jorgebucaran do you mean something similar to 
```javascript

const effect = (props, dispatch) => {
  const { target = document, name } = props;
  const event = new CustomEvent(name, props);
  target.dispatchEvent(event);
}

const Expose = {
  event: props => ({ ...props, effect }) 
}

app({
  init: { some: 'value' },
  view: () => {},
  subscriptions: state => [
    Expose.event({ name: 'stateUpdate', detail: state })
  ],
  container: document.body
})

document.addEventListener('stateUpdate', event => console.log(event.detail))
```

**Comment by sergey-shpak** [comment]
@jorgebucaran 
should we pass `state` to [sub.effects](https://github.com/jorgebucaran/hyperapp/blob/V2/src/index.js#L518) function, as `dispatch` function [does](https://github.com/jorgebucaran/hyperapp/blob/V2/src/index.js#L573) ?

**Comment by jorgebucaran** [comment]
👋 @sergey-shpak Please post questions in a new issue or join the Slack for better average response times.

**Comment by jorgebucaran** [comment]
Corrected mistakes and other small errors.

Thanks, @lukejacksonn. 😄
